File: Week 2/w2_q4.cpp

    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
           vector<int> result;
        deque<int> window;   // stores indices

        for (int i = 0; i < nums.size(); i++) {

            // Remove elements that are out of the current window
            if (!window.empty() && window.front() == i - k)
                window.pop_front();

            // Remove all smaller elements from the back
            while (!window.empty() && nums[window.back()] < nums[i])
                window.pop_back();

            window.push_back(i);

            // First complete window formed
            if (i >= k - 1)
                result.push_back(nums[window.front()]);
        }

        return result;
    }

File: Week 2/w2_q4.py
from collections import deque
from typing import List

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        deq = deque()  # stores indices
        output = []

        for i in range(len(nums)):
            # Remove indices outside current window
            while deq and deq[0] <= i - k:
                deq.popleft()

            # Remove smaller elements from the back
            while deq and nums[deq[-1]] <= nums[i]:
                deq.pop()

            deq.append(i)

            # Record answer when first window is complete
            if i >= k - 1:
                output.append(nums[deq[0]])

        return output